src/rust/protocol/src/memcache/wire/request/parse.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use super::super::*;
+use crate::*;
+
+use core::slice::Windows;
+use std::convert::TryFrom;
+
+const MAX_BYTES: usize = usize::MAX / 2;
+
+impl Parse for MemcacheRequest {
+    fn parse(buffer: &[u8]) -> Result<ParseOk<Self>, ParseError> {
+        match parse_command(buffer)? {
+            MemcacheCommand::Get => parse_get(buffer),
+            MemcacheCommand::Gets => parse_gets(buffer),
+            MemcacheCommand::Set => parse_set(buffer, false),
+            MemcacheCommand::Add => parse_add(buffer),
+            MemcacheCommand::Replace => parse_replace(buffer),
+            MemcacheCommand::Cas => parse_set(buffer, true),
+            MemcacheCommand::Delete => parse_delete(buffer),
+        }
+    }
+}
+
+struct ParseState<'a> {
+    single_byte: Windows<'a, u8>,
+    double_byte: Windows<'a, u8>,
+}
+
+impl<'a> ParseState<'a> {
+    fn new(buffer: &'a [u8]) -> Self {
+        let single_byte = buffer.windows(1);
+        let double_byte = buffer.windows(2);
+        Self {
+            single_byte,
+            double_byte,
+        }
+    }
+
+    fn next_space(&mut self) -> Option<usize> {
+        self.single_byte.position(|w| w == b" ")
+    }
+
+    fn next_crlf(&mut self) -> Option<usize> {
+        self.double_byte.position(|w| w == CRLF.as_bytes())
+    }
+}
+
+fn parse_command(buffer: &[u8]) -> Result<MemcacheCommand, ParseError> {
+    let command;
+    {
+        let mut parse_state = ParseState::new(buffer);
+        if let Some(_line_end) = parse_state.next_crlf() {
+            if let Some(cmd_end) = parse_state.next_space() {
+                command = MemcacheCommand::try_from(&buffer[0..cmd_end])?;
```

Comment:
If you want to allow `flush_all` ([ref](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L1664)) you will need to allow command conversion from either the first space or the first line.

Reply:
This now parses out single word commands - added a quick hack to implement `quit`.

Reply:
Inclined to skip flush_all command for now, let me know if you disagree on that.